### PR TITLE
fix: add api key API and middleware guards

### PR DESCRIPTION
**Pull request by tomkit**
- add endpoints to generate and revoke an api key
- add middleware to guard crm endpoints with it
- also fix some incorrect openapi examples
- guard all but top-level `/applications/:application_id` endpoints

NOTE:
- we need to add a username/password guard for mgmt endpoints for API or Management Portal access

### CODE REVIEW

File: apps/api/routes/mgmt/v1/application/index.ts [hunk 1]
```diff
@@ -87,6 +92,37 @@ export default function init(app: Router): void {
     }
   );
 
+  applicationRouter.post('/:application_id/_generate_api_key', async (req: Request, res: Response) => {
```
Comment (albertyfwu) [inline apps/api/routes/mgmt/v1/application/index.ts]: can we use the actual req and res types?

File: packages/core/types/application.ts [hunk 2]
```diff
@@ -8,7 +8,8 @@ export type Application = BaseApplication & {
 };
 
 export type ApplicationConfig = {
-  webhook: WebhookConfig | null;
+  webhook?: WebhookConfig | null;
```
Comment (albertyfwu) [inline packages/core/types/application.ts:11]: what is the reasoning for both `undefined` and `null`? same question for `apiKey`

File: apps/api/routes/mgmt/v1/application/index.ts [hunk 3]
```diff
@@ -87,9 +99,53 @@ export default function init(app: Router): void {
...
+  applicationRouter.post(
+    '/:application_id/_generate_api_key',
+    async (
+      req: Request<CreateApplicationApiKeyPathParams, CreateApplicationApiKeyResponse, CreateApplicationApiKeyRequest>,
+      res: Response<CreateApplicationApiKeyResponse>
+    ) => {
+      const application = await applicationService.getById(req.params.application_id);
+
+      const apiKey = crypto.randomBytes(64).toString('base64');
+      const hashedApiKey = await crypto.scryptSync(apiKey, SUPAGLUE_API_KEY_SALT!, 64).toString('hex'); // TODO: remove bang by getting NodeJs ProcessEnv global interface working
```
Comment (asdfryan) [inline apps/api/routes/mgmt/v1/application/index.ts:111]: We should move encryption stuff into a common lib (shared with our credentials encryption)

Comment (asdfryan) [reply apps/api/routes/mgmt/v1/application/index.ts:111]: Can move to `lib/crypt.ts`

Comment (tomkit) [reply apps/api/routes/mgmt/v1/application/index.ts:111]: Let's do as another PR? We should also consolidate the secret keys

File: openapi/mgmt/components/schemas/objects/application_api_key.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+type: object
```
Comment (asdfryan) [inline openapi/mgmt/components/schemas/objects/application_api_key.yaml:1]: We should add the required api key header too

**Review by lucasmarshall — CHANGES_REQUESTED** [review]

File: .env.sample [hunk 5]
```diff
@@ -1,7 +1,6 @@
-SUPAGLUE_API_KEY_SALT=please-change-me
-
 SUPAGLUE_SYNC_PERIOD_MS=900000
 SUPAGLUE_API_ENCRYPTION_SECRET=CHANGETHIS
+SUPAGLUE_QUICKSTART_API_KEY="8o2kGDTHwTY6zcQV8gx7ep8vV7F2JNdeu95OR75MY7dlQJhzOiM4Ha/dKwXOTfDzDxVo+a6tZBrVxYfzJFIrgw=="
```
Comment (asdfryan) [inline .env.sample:3]: I think the quotes are unnecessary

File: docker-compose.yml [hunk 6]
```diff
@@ -33,6 +33,7 @@ services:
       - SUPAGLUE_DISABLE_ERROR_REPORTING
       - SUPAGLUE_POSTHOG_API_KEY
       - SUPAGLUE_API_ENCRYPTION_SECRET
+      - SUPAGLUE_API_KEY_SALT
```
[the remark this answers is not in the document]
Comment (tomkit) [reply docker-compose.yml:36]: I consolidated into SUPAGLUE_API_ENCRYPTION_SECRET

**Review by asdfryan — APPROVED** [review]

File: docker-compose.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (lucasmarshall) [reply docker-compose.yml:36]: Not sure what you mean?

**Comment by tomkit** [comment]
landing, then we can make additional changes later

**Pull request merged**